### PR TITLE
Add configuration files for each dataset

### PR DESCRIPTION
**Pull request by AuBousquet**
Operations done in this PR:

- Delete classes XXXConfig(Config), where XXX is the name of the dataset;
- Creation of a config_XXX.yml file for each type of dataset and of a sample config.yml.sample file;
- Configuration are now read threw a dictionary, each dataset having its dictionary as a property, the configuration YAML file is passed as a parameter of the CLI, with the `-c` option;
- The `-s` parameter, giving the dataset type, is removed;
- Renaming of a few variables / parameters to improve clarity 
Examples: 
`split` giving the type of command from the CLI => `task`;  
the tasks are renamed to have `train`, `test`, `preprocess`, `validate`, `ERF`;  
`in_radius` => `config["input"]["sphere_radius"]`;  
`self.path` => `self.datapath` pour les datasets. 

- Some of the CLI parameters are now part of the config.yml files:
`--n-votes`
`--potential-increment`
`--max-epoch`
`--checkpoint-gap`
`--epoch-steps`
`--validation-size`

- Homogeneization of the datasets class structures, changing some of the variables attached to the classes
Example:
`ignored_labels` is now attached to the `PointCloudDataset` class;
`num_classes` is now a property of the class and not a property of the configuration dictionary...

NB: There is still a potential to factorize the code between the datasets.

### CODE REVIEW

**Comment by AuBousquet** [comment]
@delhomer il y a une petite régression de la couverture des tests, mais vu l'état du code, je ne vois pas bien où en rajouter utilement. Qu'en penses-tu ?

File: kpconv_torch/utils/config.py [hunk 1]
```diff
@@ -16,368 +15,29 @@ class BColors(Enum):
     UNDERLINE = "\033[4m"
 
 
-class Config:
```
Comment (delhomer) [inline kpconv_torch/utils/config.py:19]: By reading the diff, I realize that there is a misunderstanding between our comprehension of the config problem.

You choose to drop the `Config` class along with the dataset specific config classes. In my opinion, the dataset specific config classes should indeed by replaced with `.yml` files. However the main `Config` class is still useful to receive every config values and keep an object structure. (Bonus: this should have kept the diff reasonable.)

File: config.yml.sample
Comment (delhomer) [inline config.yml.sample:1]: Could you store the configuration files into a dedicated folder, for a sake of clarity?

File: tests/config_S3DIS.yml
Comment (delhomer) [inline tests/config_S3DIS.yml:1]: Could you change the name of the file, so as to indicate it is a *test* config?

File: tests/test_train.py [hunk 2]
```diff
@@ -66,61 +68,57 @@ def test_train(dataset_path, trained_model_path):
...
-        validation_size=validation_size,
     )
 
     log_dirs = list(trained_model_path.iterdir())
     assert len(log_dirs) == 1
     log_dir = log_dirs[0]
-    assert (log_dir / "parameters.txt").exists()
+
+    config = load_config(Path("tests/config_S3DIS.yml"))
+    assert (log_dir / "config.yml").exists()
```
Comment (delhomer) [inline tests/test_train.py:89]: Is `log_dir / "config.yml"` supposed to be equivalent to `tests/config_S3DIS.yml`? We should test that too.

File: tests/test_train.py [hunk 3]
```diff
@@ -66,61 +68,57 @@ def test_train(dataset_path, trained_model_path):
...
     assert (log_dir / "training.txt").exists()
     assert (log_dir / "checkpoints" / "current_chkp.tar").exists()
     assert (log_dir / "checkpoints" / "chkp_0001.tar").exists()
     training_results = np.loadtxt(log_dir / "training.txt", skiprows=1)
-    assert training_results.shape[0] == epoch_steps * max_epoch
+    assert (
+        training_results.shape[0] == config["train"]["epoch_steps"] * config["train"]["max_epoch"]
```
Comment (delhomer) [inline tests/test_train.py:95]: Use the `t` variable that you define below.

File: tests/test_train.py [hunk 4]
```diff
@@ -66,61 +68,57 @@ def test_train(dataset_path, trained_model_path):
...
-    train.train(
-        dataset_path,
-        chosen_log=log_dir,
-        output_dir=None,
-        dataset=dataset_path.name,
-    )
+    train.train(dataset_path, configfile=None, chosen_log=log_dir, output_dir=None)
 
+    log_dirs = list(trained_model_path.iterdir())
     log_dirs = list(trained_model_path.iterdir())
```
Comment (delhomer) [inline tests/test_train.py:108]: Duplicated line, any reason?

File: tests/test_train.py [hunk 5]
```diff
@@ -66,61 +68,57 @@ def test_train(dataset_path, trained_model_path):
...
     assert new_log_dir == log_dir
-    assert (log_dir / "parameters.txt").exists()
+    assert (log_dir / "config.yml").exists()
     assert (log_dir / "training.txt").exists()
     assert (log_dir / "checkpoints" / "current_chkp.tar").exists()
     assert (log_dir / "checkpoints" / "chkp_0002.tar").exists()
     training_results = np.loadtxt(log_dir / "training.txt", skiprows=1)
-    assert training_results.shape[0] == epoch_steps * (max_epoch + 1)
+    assert training_results.shape[0] == config["train"]["epoch_steps"] * (
+        config["train"]["max_epoch"] + 1
```
Comment (delhomer) [inline tests/test_train.py:118]: Use `t` ?

**Review by delhomer — CHANGES_REQUESTED** [review]

File: .github/workflows/lint.yml [hunk 6]
```diff
@@ -50,7 +50,7 @@ jobs:
 
       - name: Analysing the code with pylint
         run: |
-          pylint $(git ls-files '*.py') --fail-under 8
+          pylint $(git ls-files '*.py') --fail-under 7.8
```
Comment (delhomer) [inline .github/workflows/lint.yml:53]: Can't we increase this mark? 7.95 is not enough?

This indicator should not be the alpha and the omega of the project, however this is a guarantee to continue to develop things in a pretty satisfying direction. We can do exception (i.e. lower the required mark) to introduce big amount of code if we decide the changes are of first importance, however they should be rather limited (in frequency and intensity).

File: config_ModelNet40.yml [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  augment_color: 1.0
+  # Number of batch
+  batch_num: 10
+  # Frequency at which training checkpoints are saved on disk (in terms of epochs)
+  checkpoint_gap: 50
+  # Deformable offset loss
+  # Fitting geometry by penalizing distance from deform point to input points ('point2point')
+  # or to input point triplet ('point2plane', not implemented)
+  # Choose weights for class (used in segmentation loss). Empty list for no weights
+  class_w: {}
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply config_ModelNet40.yml]: ok, transformed as (an empty) list

File: tests/test_config.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    """
+    If the dataset is supported, valid_dataset returns it, otherwise it raises a ValueError.
+    """
+    assert valid_dataset("S3DIS") == "S3DIS"
+    with pytest.raises(ValueError):
+        valid_dataset("wrong_dataset")
+
+
+def test_load_config():
+    assert load_config("tests/config_S3DIS.yml")["dataset"] == "S3DIS"
```
Comment (delhomer) [inline tests/test_config.py:17]: You may increase (to be verified) the coverage statistics by checking the values of other config variables.

File: kpconv_torch/plot_convergence.py [hunk 9]
```diff
@@ -421,11 +416,10 @@ def compare_convergences_classif(list_of_paths, list_of_labels=None):
 
     for path in list_of_paths:
         # Load parameters
-        config = Config()
-        config.load(list_of_paths[0])
+        config = load_config(list_of_paths[0])
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/plot_convergence.py:419]: Indeed.

File: kpconv_torch/models/architectures.py [hunk 10]
```diff
@@ -268,15 +271,16 @@ def __init__(self, config, lbl_values, ign_lbls):
         self.valid_labels = np.sort([c for c in lbl_values if c not in ign_lbls])
 
         # Choose segmentation loss
-        if len(config.class_w) > 0:
-            class_w = torch.from_numpy(np.array(config.class_w, dtype=np.float32))
-            self.criterion = torch.nn.CrossEntropyLoss(weight=class_w, ignore_index=-1)
+        if len(config["train"]["class_w"]) > 0:
+            self.criterion = torch.nn.CrossEntropyLoss(
+                weight=config["train"]["class_w"], ignore_index=-1
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/models/architectures.py:276]: Indeed

File: kpconv_torch/plot_convergence.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/plot_convergence.py:419]: Indeed. I think the configurations are supposed to be all the same. As a consequence, we probably can use `path` without any loss. I did that.

File: kpconv_torch/test.py [hunk 11]
```diff
@@ -26,72 +27,38 @@
...
+    # Choose to test on validation or test task
     on_val = True
 
     # Deal with 'last_XXXXXX' choices
-    chosen_log = str(model_choice(chosen_log))
+    output_path = get_test_save_path(filename, chosen_log)
+    if configfile is not None:
+        config_file_path = configfile
+    else:
+        config_file_path = Path(chosen_log / "config.yml")
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/test.py:60]: 1. I made the choice of removing the `model_choice` function, made to find back the last trained model, in case the user asks for it from the CLI. Since the dataset is now passed to the command through the config file and not through the `-s` parameter anymore, this mechanism is obsolete. The means to find it back would be to accept, from inside the config file, to ask for `last_S3DIS` instead of `S3DIS`. As a consequence, the `-l` parameter would become useless (as it was already the cas in the former version). I did not find this way of defining the CLI very logical, so I removed this possibility and supposed that the user would ask the last one, through the correct specification of the `-l` parameter. 

2. The `if... else` structure was added following one of our discussions on a review. It is intended to distinguish the case when the inference is done with a specific config file and the case when you want to use the config file already created by the training of the model. It was absolutely necessary before I created a "test" section in the config file. Now maybe, it will be rarely used. I don't know if it can happen that the user want another config file than the one generated by the training of the model.

What do you think?

File: kpconv_torch/train.py [hunk 12]
```diff
@@ -80,174 +68,120 @@ def train(
...
-            )
-        config.chosen_log = None
-
-    # Get path from argument if given
-    if len(sys.argv) > 1:
-        config.chosen_log = sys.argv[1]
+    if chosen_log is None:
+        config_file_path = configfile
+    else:
+        config_file_path = Path(train_save_path / "config.yml")
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/train.py:75]: It is tested [here](https://github.com/delhomer/kpconv-torch/pull/57/files#diff-e34ada1abf156e2b36d35873bd3ab6976fd15b794349d0a62163ae54d972d852R129), in the CLI that there is at least `chosen_log` or `configfile` that is defined. So, it should not be possible at this step of the code. If `chosen_log` is defined, we take the configuration file from inside the folder, otherwise, we take the one defined by the CLI.

File: kpconv_torch/utils/config.py [hunk 13]
```diff
@@ -1,13 +1,12 @@
...
-import os
+import yaml
 
-import numpy as np
+from pathlib import Path
 
 
-# Colors for printing
-class BColors(Enum):
+class bcolors:
```
[the remark this answers is not in the document]
Comment (AuBousquet) [reply kpconv_torch/utils/config.py:6]: it is useless indeed

File: kpconv_torch/utils/config.py (hunk 1 above)
Comment (AuBousquet) [reply kpconv_torch/utils/config.py:19]: I think the split between the mother and the son classes has been done only following inheritance logic. That is to say, everything that is common to all datasets is in (in principle) in the mother class and the rest in the son classes. The parameters do not play a different role, depending if they are in the mother or in the son classes. That is why I decided to group them in a unique dictionary, which appears to me to be clearer than having to mechanism to query to access the parameters.

File: tests/config_S3DIS.yml
Comment (AuBousquet) [reply tests/config_S3DIS.yml:1]: yes

File: tests/test_train.py (hunk 2 above)
Comment (AuBousquet) [reply tests/test_train.py:89]: Yes it is. I added a test.

File: tests/test_train.py (hunk 4 above)
Comment (AuBousquet) [reply tests/test_train.py:113]: no

File: tests/test_train.py (hunk 3 above)
Comment (AuBousquet) [reply tests/test_train.py]: OK. `t` was defined because the use of dictionaries inside a f-string. That is why I only used it inside the f-string.

File: tests/test_config.py (hunk 8 above)
Comment (AuBousquet) [reply tests/test_config.py]: Done. We will see.

File: .github/workflows/lint.yml (hunk 6 above)
Comment (AuBousquet) [reply .github/workflows/lint.yml]: ok, fixed at 7.95.

File: kpconv_torch/test.py [hunk 14]
```diff
@@ -26,72 +27,44 @@
...
+    config = load_config(config_file_path)
+
+    try:
+        config["test"]["validation_size"]
+    except KeyError:
+        config["test"]["validation_size"] = 200
```
Comment (delhomer) [inline kpconv_torch/test.py:66]:
```suggestion
    if "validation_size" not in config["test"]:
        # consider a default value if not specified
        config["test"]["validation_size"] = 200
```

File: kpconv_torch/test.py [hunk 15]
```diff
@@ -26,72 +27,43 @@
...
+    output_path = get_test_save_path(filename, chosen_log)
+    if configfile is not None:
+        config_file_path = configfile
+    else:
+        config_file_path = Path(chosen_log / "config.yml")
+    config = load_config(config_file_path)
+
+    if "validation_size" not in config["test"]:
+        # consider a default value if not specified
+        config["test"]["validation_size"] = 200
```
Comment (delhomer) [inline kpconv_torch/test.py:65]:
```suggestion
    if "validation_size" not in config["test"]:
        # consider a default value if not specified
        config["test"]["validation_size"] = 200
```

**Comment by delhomer** [comment]
Let's consider we can ignore the coverage check there, as the diff is huge and highlights a lot of untested code.

**Review by delhomer — APPROVED** [review]
